src/cnf.cpp

```diff
@@ -0,0 +1,371 @@
+      polarities[level] = 0;
+
+      while (level-- != 0) {
+        const signed char pol = polarities[level];
+        if (pol == 0) {
+          clause_build = adapter.build_node(level, clause_build, clause_build);
+        } else {
+          polarities[level] = 0;
+          clause_build      = pol == 1 ? adapter.build_node(level, /* lo */ clause_build, tautology)
+                                       : adapter.build_node(level, /* lo */ tautology, clause_build);
+        }
+        if (level > min_var) tautology = adapter.build_node(level, tautology, tautology);
+      }
+
+      clauses.push_back(adapter.build());
+      min_var = vars.size();
+      max_var = 0;
+    }
+  }
+
+  return clauses;
+}
+
+/// Conjoin the clauses with an (approximately) balanced bracketing `(c1 ∧ c2) ∧ (c3 ∧ c4)`
+///
+/// In the paper [Configuring BDD Compilation Techniques for Feature
+/// Models](https://doi.org/10.1145/3646548.3676538), this was shown to be much
+/// better than a left-deep clause bracketing `((c1 ∧ c2) ∧ c3) ∧ c4`.
+template <typename Adapter, typename IT>
+typename Adapter::dd_t
+conjoin(Adapter& adapter, const IT begin, const IT end)
+{
+  if (begin == end) return adapter.top();
+  auto d = std::distance(begin, end);
+  if (d == 1) return *begin;
+
+  const IT mid = begin + d / 2;
+  return conjoin(adapter, begin, mid) & conjoin(adapter, mid, end);
+}
+
+// ========================================================================== //
+template <typename Adapter>
+int
+run_cnf(int argc, char** argv)
+{
+  bool should_exit = parse_input<parsing_policy>(argc, argv);
+  if (should_exit) { return -1; }
+
+  if (file.empty()) {
+    std::cerr << "Input file not specified\n";
+    return -1;
+  }
+
+  // Read the file
+  size_t nclauses;
+  std::vector<unsigned> vars_permuted;
+  std::vector<int> clauses_raw;
+
+  {
+    std::ifstream input(file, std::ios::in);
+
+    std::vector<unsigned> var_order;
+    std::string tok;
+    while (true) {
+      const int c = input.get();
+      if (c == 'c') {
+        input >> tok;
+        if (std::all_of(tok.begin(), tok.end(), ::isdigit)) {
+          const unsigned var_id = std::stoul(tok);
+          if (var_id == 0) {
+            std::cerr << "error: variable numbers must be > 0 (in variable order)\n";
+            exit(-1);
+          }
+          var_order.push_back(var_id - 1);
+        }
+        input.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
+      } else if (c == 'p') {
+        break;
+      } else {
+        std::cerr << "error: unexpected character '" << static_cast<char>(c) << "'\n";
+        exit(-1);
+      }
+    }
+
+    std::string type;
+    unsigned nvars;
+    input >> type >> nvars >> nclauses;
+
+    if (type != "cnf") {
+      std::cerr << "error: can only handle 'cnf' files\n";
+      exit(-1);
+    }
+    if (nvars >= (unsigned)std::numeric_limits<int>::max()) {
+      std::cerr << "error: too many variables\n";
+      exit(-1);
```

Comment:
I believe that there is value in notifying a user when the input exceeds the parser’s capabilities. If the BDD library has stricter requirements, that’s fine, and then the BDD library should handle the exception case accordingly. However, the parser should never silently fail and return a CNF that differs from the input file. I’m not sure if this would be undefined behavior (in the sense of the C++ language specification) if we elided the check, but exhibiting undefined behavior would be a fundamental programming mistake.

Reply:
I'm definitely using the word *undefined behaviour* very loosely here, as in: the program crashes and so what - it is just a benchmark. But, it is not a lot of extra code, so let's just keep it.